packages/x-components/src/x-modules/url/components/url-handler.vue

```diff
@@ -99,21 +110,63 @@
     /**
      * Emits the {@link UrlXEvents.ParamsLoadedFromUrl} XEvent,
      * the {@link UrlXEvents.ExtraParamsLoadedFromUrl} XEvent and, if there is query, also emits
-     * the  {@link XEventsTypes.UserOpenXProgrammatically}.
+     * the {@link XEventsTypes.UserOpenXProgrammatically}.
      *
      * @internal
      */
     protected emitEvents(): void {
       const { all, extra } = this.parseUrlParams();
-      this.$x.emit('ParamsLoadedFromUrl', all, { feature: 'url' });
-      this.$x.emit('ExtraParamsLoadedFromUrl', extra, { feature: 'url' });
+      const metadata = { ...this.createOrigin() };
+      this.$x.emit('ParamsLoadedFromUrl', all, metadata);
+      this.$x.emit('ExtraParamsLoadedFromUrl', extra, metadata);
       // TODO: Move this logic from here.
       if (all.query) {
-        this.$x.emit('UserOpenXProgrammatically');
+        this.$x.emit('UserOpenXProgrammatically', undefined, metadata);
       }
       this.urlLoaded = true;
     }
 
+    /**
+     * Creates the origin for the emitted {@link XEvent | XEvents}.
+     *
+     * @returns The metadata.
+     * @internal
+     */
+    protected createOrigin(): QueryOriginInit {
+      return {
+        feature: 'url',
+        location: this.location()
+      };
+    }
+
+    /**
+     * Retrieves the {@link FeatureLocation | location} used to build the
+     * {@link QueryOriginInit | events metadata origin}.
+     *
+     * @returns The location.
+     * @internal
+     */
+    protected location(): FeatureLocation {
```

Comment:
Rename the method name to something that has a verb in its name. `extractLocation`, `createLocation`, `detectLocation`...

Reply:
Thanks, `detectLocation` seems like a good fit.